tools/src/abstract_server/local_index.rs

```diff
@@ -0,0 +1,123 @@
+use async_trait::async_trait;
+use flate2::read::GzDecoder;
+use serde_json::{from_str, Value};
+use std::io::Read;
+use tokio::fs::File;
+use tokio::io::AsyncReadExt;
+
+use super::server_interface::{AbstractServer, ErrorDetails, ErrorLayer, Result, ServerError};
+
+use crate::config::{load, TreeConfigPaths};
+
+/// IO errors amount to a 404 for our purposes which means a sticky problem.
+impl From<std::io::Error> for ServerError {
+    fn from(err: std::io::Error) -> ServerError {
+        ServerError::StickyProblem(ErrorDetails {
+            layer: ErrorLayer::ServerLayer,
+            message: err.to_string(),
+        })
+    }
+}
+
+/// Read newline-delimited JSON that's been gzip-compressed.
+async fn read_gzipped_ndjson_from_file(path: &str) -> Result<Vec<Value>> {
+    let mut f = File::open(path).await?;
+    // We read the entirety to a buffer because
+    // https://github.com/serde-rs/json/issues/160 suggests that the buffered
+    // reader performance is likely to be much worse.
+    //
+    // Alternately, the flate2 library can be built with the "tokio" feature and
+    // we could do fully buffered decoding.
+    let mut buffer = Vec::new();
+    f.read_to_end(&mut buffer).await?;
+
+    let mut gz = GzDecoder::new(&buffer[..]);
+
+    let mut raw_str = String::new();
+    gz.read_to_string(&mut raw_str)?;
+
+    // let mut raw_str = String::new();
+    // f.read_to_string(&mut raw_str).await?;
+
+    Ok(raw_str.lines().filter_map(|s| from_str(s).ok()).collect())
+}
+
+#[derive(Debug)]
+struct LocalIndex {
+    // We only hold onto the TreeConfigPaths portion of the config because the
+    // git data is not `Sync`.  Specifically, the compiler says:
+    //
+    // "within `TreeConfig`, the trait `Sync` is not implemented for
+    // `*mut libgit2_sys::git_repository`"
+    //
+    // Obviously, it would be nice to be able to do git stuff, but that needs to
```

Comment:
I've transformed my hand-waving comments here to basically what you've said here, cross-referencing the `buld-blame.rs` compute threads.  One implication here that I didn't leave in the comment is that presumably this means that all the git stuff will need to be done in straight runs that never awaits (because that allows thread changes and the type signatures enforce this).